### ISSUE TITLE
Bug: experiment.py continuous_train_and_eval() leads to overfitting

### ISSUE THREAD

**Issue by dagarcia-nvidia** [tensorflow/tensorflow#11013]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: YES
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 14.04
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: v1.0.1-4-gbd6743e-dirty 1.0.1
- **Bazel version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem
Our team has found what we believe is a critical bug in [experiment.py continuous_train_and_eval()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/experiment.py#L506) that pretty much guarantees any model trained with this method will lead to overfitting.

The problem is that every time it alternates from evaluating to training again, it calls [self._call_train()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/experiment.py#L569), which eventually makes it to [estimator.py _train_model()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L950).

This function first [creates a brand new graph](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L952) then [sets up the input pipelines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L956). This necessarily means that the input pipelines are reset from scratch on every training lap, and since a training lap is typically much shorter than an epoch, this means the model is only trained with a small portion of the whole dataset.

Originally we raised this as an issue in [Google's tf-seq2seq](https://github.com/google/seq2seq/issues/262), but as far as we can tell now the bug probably belongs into Tensorflow instead, as we can't see a simple way to modify the tf-seq2seq code to avoid this issue from arising.

Thank you for looking into this matter.

### Source code / logs
N/A

**Comment by tatatodd** [comment]
@jhseu Can you comment on this?  Thanks!

**Comment by jhseu** [comment]
Yeah, that's true. Often input_fns are defined to randomize inputs, but this does seem like an issue. @ispirmustafa @xiejw

**Comment by dagarcia-nvidia** [comment]
I see. In the [original issue](https://github.com/google/seq2seq/issues/262) the client was calling [tf.contrib.slim.dataset.Dataset()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/data/dataset.py#L38) which doesn't appear to have an option to thoroughly randomize its inputs.

I'm pretty new to TF, so I may be missing lots of things here :)

**Comment by xiejw** [comment]
Shuffle=True for train input_fn is recommended. And also it is best to make the training steps for each iteration longer, similar to the epoch size or multiple epochs. I will fix the docstring to document this.

**Comment by xiejw** [comment]
not familiar with the slim Dataset api, but according to the docstring, there is a provider, which support randomize.

**Comment by dagarcia-nvidia** [comment]
@xiejw,

Setting shuffle=true in a tf.RandomShuffleQueue would hide the problem a bit, but not eliminate it, because the "random" part of a random queue occurs [when dequeuing](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/random_shuffle_queue_op.cc#L116). Since queues have a limited size, and the state is reset on every training lap, the model would still not train with the whole dataset.

It seems like the way to fix this properly is by not regenerating the graph from scratch on every training / testing lap. The graphs should be generated only once and then reused.

**Comment by xiejw** [comment]
Experiment has a method train_and_evaluate, which does the approach you wanted. However, it has a double memory problem, which is why continous_train_and_evaluate was introduced (as experimental). Can you try make train steps for each iteration larger (so one train iteration can process one epoch training data or multiple epoch data)?  

If that does not solve the problem, you probably need to try train_and_evaluate or write a hook to do evaluation by calling train().

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.
